ember-core/src/main/scala/org/http4s/ember/core/Parser.scala

```diff
@@ -30,13 +30,17 @@ private[ember] object Parser {
 
   object HeaderP {
 
-    def parseHeaders[F[_]: MonadThrow](
-        s: Stream[F, Byte],
+    def parseHeaders[F[_]](
+        head: Array[Byte],
+        read: F[Option[Chunk[Byte]]],
         maxHeaderLength: Int,
-        acc: Option[ParseHeadersIncomplete])
-        : Pull[F, Nothing, (Headers, Boolean, Option[Long], Stream[F, Byte])] =
-      s.pull.uncons.flatMap {
-        case Some((chunk, tl)) =>
+        acc: Option[ParseHeadersIncomplete])(implicit
+        F: MonadThrow[F]): F[(Headers, Boolean, Option[Long], Array[Byte])] = {
+      // TODO: improve this
+      val uncons = if (head.nonEmpty) F.pure(Some(Chunk.bytes(head))) else read
```

Comment:
This pattern basically shows up everywhere. I wonder if we can stick `head` and `read` together to form a more interesting type